### PR TITLE
added transitive reduction algorithm and unit test

### PR DESCRIPTION
**Pull request by cthiebaud**
this is a port from python sample code by stackoverflow user Michael Clerx, posted as an answer to question http://stackoverflow.com/questions/1690953/transitive-reduction-algorithm-pseudocode.

maybe it is ok to include in jgrapht core, thanks to review my pull request.

Christophe

### CODE REVIEW

**Comment by cthiebaud** [comment]
FYI, I used this algorithm to tweak the display of generated dependency graphs such as

image

the red arrow points to a link which is removed from the jgrapht dependency graph after transitive reduction.

File: jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * 
+ */
+public class TransitiveReduction<V, E> {
+
+    final private List<V> vertices;
+    final private int [][] pathMatrix;
+    
+    private final DirectedGraph<V, E> graph;
+    
+    public TransitiveReduction(DirectedGraph<V, E> graph) {
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java]: Move all of the work from the constructor to the reduce call.

File: jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
+package org.jgrapht.alg;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java]: Add standard JGraphT file headers.

File: jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import org.jgrapht.DirectedGraph;
+
+/**
+ *
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java]: Use Javadoc html conventions; see DijkstraShortestPath for an example.

File: jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        
+        int n = pathMatrix.length;
+        int[][] transitivelyReducedMatrix = new int[n][n];
+        System.arraycopy(pathMatrix, 0, transitivelyReducedMatrix, 0, pathMatrix.length);
+        transitiveReduction(transitivelyReducedMatrix);
+        
+        for (int i = 0; i <n; i++) {
+            for (int j = 0; j < n; j++) { 
+                if (transitivelyReducedMatrix[i][j] == 0) {
+                    // System.out.println("removing "+vertices.get(i)+" -> "+vertices.get(j));
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/TransitiveReduction.java]: Remove all commented-out printlns.

**Comment by jsichi** [comment]
Thanks; a few minor things need to be cleaned up and then this can be included into JGraphT.

We're starting to see a number of matrix representations creep into individual algorithms, so in the future maybe someone will come along and consolidate them into a reusable package.

**Comment by cthiebaud** [comment]
I only see now your comments. In the meanwhile, I have carelessly deleted my original fork on github, so it seems I have to create a new pull request, containing this original one + your recommendations. I'll call it "added transitive reduction algorithm and unit test 2". I did also changed my code to use BitSet, to hopefully reduce the memory footprint.

**Pull request closed without merging**